### PR TITLE
Slasher highest source target

### PR DESCRIPTION
**Pull request by alonmuroch**
The goal of this work is to add to the slasher the ability to track highest source/ target for every validator and then expose it via API.
This is all part of minimal slashing protection for validators restoring only from seed (without their attestation history).

Conversation: https://discordapp.com/channels/476244492043812875/730447046930071672/750732958054219806

### CODE REVIEW

**Comment by alonmuroch** [comment]
@shayzluf @terencechain would appreciate any feedback if this direction is good to continue the PR

**Comment by terencechain** [comment]
> @shayzluf @terencechain would appreciate any feedback if this direction is good to continue the PR

I'm not deep in the weeds with the slasher architecture. Should we write up design doc with this? It makes it easier to explore the trade offs and evaluate the necessity of this

**Comment by alonmuroch** [comment]
@terencechain you mean for adding the highest att. or in general for the slasher?

btw I've mostly used the existing arch. using the cache already used.

File: slasher/db/kv/highest_attestation.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+	"go.opencensus.io/trace"
+)
+
+func persistHighestAttestationCacheOnEviction(db *Store) func(key interface{}, value interface{}) {
+	// We use a closure here so we can access the database itself
+	// on the eviction of a span map from the cache. The function has the signature
+	// required by the ristretto cache OnEvict method.
+	// See https://godoc.org/github.com/dgraph-io/ristretto#Config.
+	return func(key interface{}, value interface{}) {
+
```
Comment (shayzluf) [inline slasher/db/kv/highest_attestation.go:19]: if you don't write on eviction you never write cache into db

File: slasher/db/kv/highest_attestation.go [hunk 2]
```diff
@@ -0,0 +1,81 @@
...
+		db.highestAttestationCache.Set(validatorID, highest)
+		return nil
+	}
+
+	enc, err := json.Marshal(highest)
+	if err != nil {
+		return errors.Wrap(err, "failed to marshal")
+	}
+
+	key := []byte(fmt.Sprintf("%d", validatorID))
```
Comment (shayzluf) [inline slasher/db/kv/highest_attestation.go:68]: we use `bytesutil.Uint64ToBytesBigEndian()` to have the keys sorted and binary searchable

File: slasher/detection/detect.go [hunk 3]
```diff
@@ -216,3 +216,37 @@ func isSurrounding(incomingAtt *ethpb.IndexedAttestation, prevAtt *ethpb.Indexed
...
+			h.HighestSourceEpoch = att.Data.Source.Epoch
+			update = true
+		}
+		if h.HighestTargetEpoch < att.Data.Target.Epoch {
+			h.HighestTargetEpoch = att.Data.Target.Epoch
+			update = true
+		}
+
+		if update {
+			if err := ds.slasherDB.SaveHighestAttestation(ctx, idx, h); err != nil {
```
Comment (shayzluf) [inline slasher/detection/detect.go:243]: what if highest target is higher but source is lower . do we want to update highest source to a lower value? (same for the opposite)

File: slasher/db/kv/highest_attestation.go (hunk 1 above)
Comment (alonmuroch) [reply slasher/db/kv/highest_attestation.go:19]: yep I know, not implemented yet.

File: slasher/detection/service.go [hunk 4]
```diff
@@ -196,6 +196,11 @@ func (ds *Service) detectHistoricalChainData(ctx context.Context) {
 				}
 			}
 			ds.submitAttesterSlashings(ctx, slashings)
+
+			// update highest attestation
+			if err := ds.UpdateHighestAttestation(ctx, att); err != nil {
```
Comment (shayzluf) [inline slasher/detection/service.go:201]: you should also update highest on isslashble attestation endpoint

File: slasher/db/kv/highest_attestation.go (hunk 2 above)
Comment (alonmuroch) [reply slasher/db/kv/highest_attestation.go:68]: got it

File: slasher/detection/detect.go (hunk 3 above)
Comment (alonmuroch) [reply slasher/detection/detect.go:243]: can that even happen without slashing? have an example?

File: slasher/detection/service.go (hunk 4 above)
Comment (alonmuroch) [reply slasher/detection/service.go:201]: gotcha

**Comment by alonmuroch** [comment]
@shayzluf is it safe to say this implementation direction is good enough?
Do we have any way to test if it affects performance?

**Comment by shayzluf** [comment]
@alonmuroch The direction seems fine, writing and reading from cache wont effect  the performance other then have memory requirements implications. There should be another pr to add an endpoint that query for these values

**Comment by alonmuroch** [comment]
> @alonmuroch The direction seems fine, writing and reading from cache wont effect the performance other then have memory requirements implications. There should be another pr to add an endpoint that query for these values

Ok so I'll complete this PR and start running it alongside our infra to test and see if mem requirements jump.

File: slasher/detection/detect.go (hunk 3 above)
Comment (alonmuroch) [reply slasher/detection/detect.go:243]: @shayzluf I think in that case it's a slashable attestation

File: slasher/detection/service.go (hunk 4 above)
Comment (alonmuroch) [reply slasher/detection/service.go:201]: you mean return highest there?

**Comment by alonmuroch** [comment]
@shayzluf @terencechain updated to a full PR

File: slasher/cache/highest_attestation_cache.go [hunk 5]
```diff
@@ -0,0 +1,62 @@
+package cache
+
+import (
+	lru "github.com/hashicorp/golang-lru"
+	slashpb "github.com/prysmaticlabs/prysm/proto/slashing"
+)
+
+var (
+	// validatorsCacheSize defines the max number of validators public keys the cache can hold.
+	highestAttCacheSize = 100000
```
Comment (shayzluf) [inline slasher/cache/highest_attestation_cache.go:10]: i think 300000 is better as we don't want to write and read from db

Comment (alonmuroch) [reply slasher/cache/highest_attestation_cache.go]: @shayzluf done

File: slasher/detection/detect.go (hunk 3 above)
Comment (shayzluf) [reply slasher/detection/detect.go]: it is slashable but i think slasher protection should still protect it on the best way (might not be caught as slashingand will be slasher by the later offence)

File: proto/slashing/slashing.proto [hunk 6]
```diff
@@ -25,6 +25,23 @@ service Slasher {
     // Returns if a given beacon block header could be slashable when compared to the slashers history for the proposer.
     // This function is read-only, and does not need the beacon block header to be signed.
     rpc IsSlashableBlockNoUpdate(ethereum.eth.v1alpha1.BeaconBlockHeader) returns (Slashable);
+
```
Comment (shayzluf) [inline proto/slashing/slashing.proto:28]: can we separate the endpoint impl to another pr?

File: slasher/detection/service.go (hunk 4 above)
Comment (shayzluf) [reply slasher/detection/service.go:201]: update highest if it changes through the endpoint

File: proto/slashing/slashing.proto (hunk 6 above)
Comment (alonmuroch) [reply proto/slashing/slashing.proto]: @shayzluf done

**Comment by alonmuroch** [comment]
@terencechain @shayzluf add linter, build and test fixes.
Also, found out it takes forever to persist the cache data so I changed the way it saves it.

File: slasher/cache/highest_attestation_cache.go [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+
+import (
+	lru "github.com/hashicorp/golang-lru"
+	slashpb "github.com/prysmaticlabs/prysm/proto/slashing"
+	log "github.com/sirupsen/logrus"
+)
+
+var (
+	// highestAttCacheSize defines the max number of sets of highest attestation in cache.
+	highestAttCacheSize = 3000
```
Comment (shayzluf) [inline slasher/cache/highest_attestation_cache.go:11]:
```suggestion
	highestAttCacheSize = 300000
```

File: slasher/cache/highest_attestation_cache.go [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+	return c.cache.Remove(setKey)
+}
+
+// Has returns true if the key exists in the cache.
+func (c *HighestAttestationCache) Has(setKey uint64) bool {
+	return c.cache.Contains(setKey)
+}
+
+// Clear removes all keys from the ValidatorCache.
+func (c *HighestAttestationCache) Clear() {
```
Comment (shayzluf) [inline slasher/cache/highest_attestation_cache.go:65]: whats the difference between clear and purge

File: slasher/db/kv/highest_attestation.go [hunk 9]
```diff
@@ -0,0 +1,128 @@
...
+
+	return nil
+}
+
+func highestAttSetkeyBytes(validatorID uint64) []byte {
+	return bytesutil.Uint64ToBytesBigEndian(highestAttSetkey(validatorID))
+}
+
+func highestAttSetkey(validatorID uint64) uint64 {
+	return validatorID % 1000
```
Comment (shayzluf) [inline slasher/db/kv/highest_attestation.go:127]: can you please explain why are you using % here?

File: slasher/db/kv/schema.go [hunk 10]
```diff
@@ -16,6 +16,7 @@ var (
 	// Slasher-related buckets.
 	historicIndexedAttestationsBucket = []byte("historic-indexed-attestations-bucket")
 	historicBlockHeadersBucket        = []byte("historic-block-headers-bucket")
+	highestAttestationBucket          = []byte("historic-block-headers-bucket")
```
Comment (shayzluf) [inline slasher/db/kv/schema.go:19]: running over the same bucket name

File: slasher/flags/flags.go [hunk 11]
```diff
@@ -57,4 +57,11 @@ var (
 		Usage: "Sets the span cache size.",
 		Value: 1500,
 	}
+	// HighestAttCacheSize is a flag that sets the size of span cache.
+	HighestAttCacheSize = &cli.IntFlag{
+		Name:  "highest-att-cache-size",
+		Usage: "Sets the highest attestation cache size.",
+		Value: 3000,
```
Comment (shayzluf) [inline slasher/flags/flags.go:64]:
```suggestion
		Value: 300000,
```

File: slasher/detection/detect_test.go [hunk 12]
```diff
@@ -349,6 +350,146 @@ func TestDetect_detectAttesterSlashings_Double(t *testing.T) {
...
+			name: "update only target to higher",
+			savedHighest: &ethereum_slashing.HighestAttestation{
+				ValidatorId:          1,
+				HighestSourceEpoch:   1,
+				HighestTargetEpoch:   2,
+			},
+			incomingAtt: &ethpb.IndexedAttestation{
+				AttestingIndices: []uint64{1, 3, 7},
+				Data: &ethpb.AttestationData{
+					Source:          &ethpb.Checkpoint{Epoch: 2, Root: make([]byte, 32)},
```
Comment (shayzluf) [inline slasher/detection/detect_test.go:370]:
```suggestion
					Source:          &ethpb.Checkpoint{Epoch: 1, Root: make([]byte, 32)},
```

File: slasher/detection/detect_test.go [hunk 13]
```diff
@@ -349,6 +350,146 @@ func TestDetect_detectAttesterSlashings_Double(t *testing.T) {
...
+					Target:          &ethpb.Checkpoint{Epoch: 4, Root: make([]byte, 32)},
+					BeaconBlockRoot: make([]byte, 32),
+				},
+				Signature: bytesutil.PadTo([]byte{1, 2}, 96),
+			},
+			expected: &ethereum_slashing.HighestAttestation{
+				ValidatorId:          1,
+				HighestSourceEpoch:   2,
```
Comment (shayzluf) [inline slasher/detection/detect_test.go:378]:
```suggestion
				HighestSourceEpoch:   1,
```

File: slasher/cache/highest_attestation_cache.go (hunk 7 above)
Comment (alonmuroch) [reply slasher/cache/highest_attestation_cache.go:11]: no need for so much, the attestations are saved in 1k buckets

File: slasher/db/kv/highest_attestation.go (hunk 9 above)
Comment (alonmuroch) [reply slasher/db/kv/highest_attestation.go:127]: Yeah that isn't written well, basic idea is to save attestations in buckets base on validator id to optimise when persisting to db.
I'll fix that.

File: slasher/cache/highest_attestation_cache.go (hunk 8 above)
Comment (alonmuroch) [reply slasher/cache/highest_attestation_cache.go:65]: actually followed the other cache implementations here..

File: slasher/db/kv/schema.go (hunk 10 above)
Comment (alonmuroch) [reply slasher/db/kv/schema.go:19]: done

File: slasher/flags/flags.go (hunk 11 above)
Comment (alonmuroch) [reply slasher/flags/flags.go:64]: no need for it to be so big as it's saved in 1k buckets

**Comment by alonmuroch** [comment]
@shayzluf done

File: proto/slashing/slashing.pb.go [hunk 14]
```diff
@@ -6,17 +6,16 @@ package ethereum_slashing
...
 	_ "github.com/gogo/protobuf/gogoproto"
 	proto "github.com/gogo/protobuf/proto"
 	v1alpha1 "github.com/prysmaticlabs/ethereumapis/eth/v1alpha1"
 	github_com_prysmaticlabs_go_bitfield "github.com/prysmaticlabs/go-bitfield"
 	grpc "google.golang.org/grpc"
 	codes "google.golang.org/grpc/codes"
 	status "google.golang.org/grpc/status"
+	io "io"
+	math "math"
+	math_bits "math/bits"
```
Comment (shayzluf) [inline proto/slashing/slashing.pb.go:18]: please run `goimports -w .`

File: slasher/db/kv/highest_attestation.go [hunk 15]
```diff
@@ -0,0 +1,129 @@
+package kv
+
+import (
+	"context"
+	"encoding/json"
+	"github.com/pkg/errors"
```
Comment (shayzluf) [inline slasher/db/kv/highest_attestation.go:6]: goimports

**Comment by shayzluf** [comment]
@alonmuroch can you please run `goimports -w . `

**Comment by alonmuroch** [comment]
@shayzluf done

**Comment by shayzluf** [comment]
@alonmuroch still some gofmt issues 
```
Following files need to be properly formatted:
beacon-chain/main.go
slasher/main.go
validator/main.go
```

**Comment by alonmuroch** [comment]
@shayzluf that's out of scope for my PR I think... those files are not part of the slasher. 
I don't mind running gofmt on them but I think we should maybe separate it.

**Comment by alonmuroch** [comment]
@terencechain @shayzluf I'd really like to finish up this PR.. what's next?

**Comment by shayzluf** [comment]
can you fix the conflicts and make it pass tests today. i will take or of getting another review tomorrow

File: slasher/db/iface/interface.go [hunk 16]
```diff
@@ -7,6 +7,8 @@ import (
 	"context"
 	"io"
 
+	slashpb "github.com/prysmaticlabs/prysm/proto/slashing"
+
```
Comment (shayzluf) [inline slasher/db/iface/interface.go:11]:
```suggestion
```

File: slasher/db/kv/highest_attestation.go [hunk 17]
```diff
@@ -0,0 +1,131 @@
...
+	// required by the ristretto cache OnEvict method.
+	// See https://godoc.org/github.com/dgraph-io/ristretto#Config.
+	return func(key interface{}, value interface{}) {
+		log.Tracef("Evicting highest attestation for validator: %d", key.(uint64))
+		err := db.update(func(tx *bolt.Tx) error {
+			enc, err := json.Marshal(value.(map[uint64]*slashpb.HighestAttestation))
+			if err != nil {
+				return errors.Wrap(err, "failed to marshal")
+			}
+
```
Comment (shayzluf) [inline slasher/db/kv/highest_attestation.go:27]:
```suggestion
```

File: slasher/db/kv/highest_attestation.go [hunk 18]
```diff
@@ -0,0 +1,131 @@
...
+			dbKey := highestAttSetkeyBytes(key.(uint64))
+
```
Comment (shayzluf) [inline slasher/db/kv/highest_attestation.go:29]:
```suggestion
```

File: slasher/detection/detect.go [hunk 19]
```diff
@@ -4,6 +4,8 @@ import (
 	"bytes"
 	"context"
 
+	slashpb "github.com/prysmaticlabs/prysm/proto/slashing"
+
```
Comment (shayzluf) [inline slasher/detection/detect.go:8]:
```suggestion
```

File: slasher/detection/detect_test.go [hunk 20]
```diff
@@ -5,6 +5,8 @@ import (
 	"reflect"
 	"testing"
 
+	ethereum_slashing "github.com/prysmaticlabs/prysm/proto/slashing"
+
```
Comment (shayzluf) [inline slasher/detection/detect_test.go:9]:
```suggestion
```

File: slasher/detection/listeners.go [hunk 21]
```diff
@@ -70,6 +70,11 @@ func (ds *Service) detectIncomingAttestations(ctx context.Context, ch chan *ethp
 				}
 			}
 			ds.submitAttesterSlashings(ctx, slashings)
+
+			// update highest attestation
```
Comment (shayzluf) [inline slasher/detection/listeners.go:74]:
```suggestion
```

File: slasher/detection/service.go [hunk 22]
```diff
@@ -196,6 +196,11 @@ func (ds *Service) detectHistoricalChainData(ctx context.Context) {
 				}
 			}
 			ds.submitAttesterSlashings(ctx, slashings)
+
+			//update highest attestation
```
Comment (shayzluf) [inline slasher/detection/service.go:200]:
```suggestion
```

File: slasher/flags/flags.go [hunk 23]
```diff
@@ -57,4 +57,10 @@ var (
 		Usage: "Sets the span cache size.",
 		Value: 1500,
 	}
+	// HighestAttCacheSize is a flag that sets the size of span cache.
```
Comment (shayzluf) [inline slasher/flags/flags.go:60]:
```suggestion
	// HighestAttCacheSize is a flag that sets the size of highest attestation cache.
```

File: slasher/detection/service.go [hunk 24]
```diff
@@ -196,6 +196,10 @@ func (ds *Service) detectHistoricalChainData(ctx context.Context) {
 				}
 			}
 			ds.submitAttesterSlashings(ctx, slashings)
+
+			if err := ds.UpdateHighestAttestation(ctx, att); err != nil {
+				log.WithError(err).Errorf("could not update highest attestation")
+			}
 		}
 		latestStoredHead = &ethpb.ChainHead{HeadEpoch: epoch}
 		if err := ds.slasherDB.SaveChainHead(ctx, latestStoredHead); err != nil {
```
Comment (farazdagi) [inline slasher/detection/service.go:205]: Code seems to be untested (no unit tests covering it):
image

(red lines denote uncovered code)

**Review by farazdagi — CHANGES_REQUESTED** [review]
Very solid work, thank you!

My review is not about the implemented feature itself, but more about overall coding style. So, nothing huge is spotted, just some suggestions that should make code easier to understand and in line with what we already have.

File: slasher/detection/service.go (hunk 24 above)
Comment (alonmuroch) [reply slasher/detection/service.go:205]: Not sure any of function in service.go are tested..

**Comment by alonmuroch** [comment]
@shayzluf @farazdagi pushed fixes to all comments

**Review by shayzluf — COMMENTED** [review]
lgtm. please fix the conflict @alonmuroch

**Comment by rauljordan** [comment]
Hi @alonmuroch just a friendly reminder to resolve conflicts

**Comment by alonmuroch** [comment]
@rauljordan I did but for some reason GitHub marked our repository for this PR as unknown so the latest fix wasn't propagated. 
Please check out the PR (exactly the same just a new one): https://github.com/prysmaticlabs/prysm/pull/7604#pullrequestreview-515079822
